The job **`pr-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/20990980023/job/60336234397) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/60336234397) [(plain)](https://github.com/rust-lang/rust/commit/14017fc9537a1f0acdfa8db4713e8f9557534c87/checks/60336234397/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   |
69 | extern crate proc_macro2;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^

error: only metadata stub found for `rlib` dependency `quote`
       please provide path to the corresponding .rmeta file with full metadata
  --> /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/serde_derive-1.0.228/src/lib.rs:70:1
   |
70 | extern crate quote;
   | ^^^^^^^^^^^^^^^^^^^

error: only metadata stub found for `rlib` dependency `syn`
       please provide path to the corresponding .rmeta file with full metadata
  --> /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/serde_derive-1.0.228/src/lib.rs:71:1
   |
71 | extern crate syn;
   | ^^^^^^^^^^^^^^^^^

error[E0432]: unresolved imports `quote::quote_spanned`, `quote::ToTokens`, `quote::quote_spanned`, `quote::quote_spanned`, `quote::quote_spanned`, `quote::ToTokens`, `quote::quote_spanned`, `quote::quote_spanned`, `quote::format_ident`, `quote::quote_spanned`
  --> /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/serde_derive-1.0.228/src/de.rs:8:20
   |
 8 | use quote::{quote, quote_spanned, ToTokens};
   |                    ^^^^^^^^^^^^^  ^^^^^^^^
   |
  ::: /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/serde_derive-1.0.228/src/de/enum_adjacently.rs:15:20
   |
15 | use quote::{quote, quote_spanned};
   |                    ^^^^^^^^^^^^^
   |
  ::: /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/serde_derive-1.0.228/src/de/enum_externally.rs:19:20
   |
19 | use quote::{quote, quote_spanned};
   |                    ^^^^^^^^^^^^^
   |
  ::: /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/serde_derive-1.0.228/src/de/enum_untagged.rs:18:20
   |
18 | use quote::{quote, quote_spanned};
   |                    ^^^^^^^^^^^^^
   |
  ::: /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/serde_derive-1.0.228/src/de/identifier.rs:10:20
   |
10 | use quote::{quote, ToTokens};
   |                    ^^^^^^^^
   |
  ::: /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/serde_derive-1.0.228/src/de/struct_.rs:13:20
   |
13 | use quote::{quote, quote_spanned};
   |                    ^^^^^^^^^^^^^
   |
  ::: /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/serde_derive-1.0.228/src/de/tuple.rs:9:20
   |
 9 | use quote::{quote, quote_spanned};
   |                    ^^^^^^^^^^^^^
   |
  ::: /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/serde_derive-1.0.228/src/pretend.rs:4:13
   |
 4 | use quote::{format_ident, quote};
   |             ^^^^^^^^^^^^
   |
  ::: /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/serde_derive-1.0.228/src/ser.rs:8:20
   |
 8 | use quote::{quote, quote_spanned};
   |                    ^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0432`.
[RUSTC-TIMING] serde_derive test:false 0.125
error: could not compile `serde_derive` (lib) due to 4 previous errors
```

</details>

--- a/src/bootstrap/src/core/build_steps/compile.rs
+++ b/src/bootstrap/src/core/build_steps/compile.rs
@@ -2660,7 +2660,7 @@ pub enum ArtifactKeepMode {
 
 pub fn run_cargo(
     builder: &Builder<'_>,
-    cargo: Cargo,
+    mut cargo: Cargo,
     tail_args: Vec<String>,
     stamp: &BuildStamp,
     additional_target_deps: Vec<(PathBuf, DependencyType)>,
@@ -2677,6 +2677,16 @@ pub fn run_cargo(
         .parent()
         .unwrap() // chop off `$target`
         .join(target_root_dir.file_name().unwrap());
+
+    match artifact_keep_mode {
+        ArtifactKeepMode::OnlyDylib
+        | ArtifactKeepMode::OnlyRmeta
+        | ArtifactKeepMode::BothRlibAndRmeta
+        | ArtifactKeepMode::Custom(_) => {
+            cargo.arg("-Zno-embed-metadata");
+        }
+        ArtifactKeepMode::OnlyRlib => {}
+    }
 
     // Spawn Cargo slurping up its JSON output. We'll start building up the
     // `deps` array of all files it generated along with a `toplevel` array of

--- a/src/bootstrap/src/core/build_steps/compile.rs
+++ b/src/bootstrap/src/core/build_steps/compile.rs
@@ -1185,13 +1185,15 @@ impl Step for Rustc {
                 {
                     // jemalloc_sys and rustc_public_bridge are not linked into librustc_driver.so,
                     // so we need to distribute them as rlib to be able to use them.
-                    filename.ends_with(".rlib")
-                } else {
-                    // Distribute the rest of the rustc crates as rmeta files only to reduce
-                    // the tarball sizes by about 50%. The object files are linked into
-                    // librustc_driver.so, so it is still possible to link against them.
-                    filename.ends_with(".rmeta")
+                    if filename.ends_with(".rlib") {
+                        return true;
+                    }
                 }
+
+                // Distribute the rest of the rustc crates as rmeta files only to reduce
+                // the tarball sizes by about 50%. The object files are linked into
+                // librustc_driver.so, so it is still possible to link against them.
+                filename.ends_with(".rmeta")
             })),
         );
 
@@ -1741,7 +1743,7 @@ impl Step for GccCodegenBackend {
 
         let _guard =
             builder.msg(Kind::Build, "codegen backend gcc", Mode::Codegen, build_compiler, host);
-        let files = run_cargo(builder, cargo, vec![], &stamp, vec![], ArtifactKeepMode::OnlyRlib);
+        let files = run_cargo(builder, cargo, vec![], &stamp, vec![], ArtifactKeepMode::OnlyDylib);
 
         GccCodegenBackendOutput {
             stamp: write_codegen_backend_stamp(stamp, files, builder.config.dry_run()),
@@ -1817,7 +1819,7 @@ impl Step for CraneliftCodegenBackend {
             build_compiler,
             target,
         );
-        let files = run_cargo(builder, cargo, vec![], &stamp, vec![], ArtifactKeepMode::OnlyRlib);
+        let files = run_cargo(builder, cargo, vec![], &stamp, vec![], ArtifactKeepMode::OnlyDylib);
         write_codegen_backend_stamp(stamp, files, builder.config.dry_run())
     }
 
@@ -2641,6 +2643,8 @@ pub fn add_to_sysroot(
 /// build stamp, and thus be included in dist archives and copied into sysroots by default.
 /// Note that some kinds of artifacts are copied automatically (e.g. native libraries).
 pub enum ArtifactKeepMode {
+    /// Only keep .so files, ignore .rlib and .rmeta files
+    OnlyDylib,
     /// Only keep .rlib files, ignore .rmeta files
     OnlyRlib,
     /// Only keep .rmeta files, ignore .rlib files
@@ -2703,6 +2707,7 @@ pub fn run_cargo(
                 true
             } else {
                 match &artifact_keep_mode {
+                    ArtifactKeepMode::OnlyDylib => false,
                     ArtifactKeepMode::OnlyRlib => filename.ends_with(".rlib"),
                     ArtifactKeepMode::OnlyRmeta => filename.ends_with(".rmeta"),
                     ArtifactKeepMode::BothRlibAndRmeta => {

--- a/src/bootstrap/src/core/builder/cargo.rs
+++ b/src/bootstrap/src/core/builder/cargo.rs
@@ -1095,10 +1095,6 @@ impl Builder<'_> {
         // Enable usage of unstable features
         cargo.env("RUSTC_BOOTSTRAP", "1");
 
-        if matches!(mode, Mode::Std | Mode::Rustc | Mode::Codegen) {
-            cargo.arg("-Zno-embed-metadata");
-        }
-
         if self.config.dump_bootstrap_shims {
             prepare_behaviour_dump_dir(self.build);
 

--- a/src/bootstrap/src/core/builder/cargo.rs
+++ b/src/bootstrap/src/core/builder/cargo.rs
@@ -1095,7 +1095,7 @@ impl Builder<'_> {
         // Enable usage of unstable features
         cargo.env("RUSTC_BOOTSTRAP", "1");
 
-        if matches!(mode, Mode::Std) {
+        if matches!(mode, Mode::Std | Mode::Rustc | Mode::Codegen) {
             cargo.arg("-Zno-embed-metadata");
         }
 

